Add Flexit bacnet integration

lellky opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
New integration for Flexit Nordic air handling units using built-in BACnet support.
Existing 'flexit' integration requires an additional Modbus adapter installed, so this is simpler and cheaper to use.

This is a new try of getting the Flexit Integration into Home Assistant. Previous try: https://github.com/home-assistant/core/pull/90006

All cred for writing the integration and the underlying library goes to @piotrbulinski .

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29939

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

piotrbulinski committed:
Add Flexit (BACnet) integration

piotrbulinski committed:
Add missing test for config_flow

piotrbulinski committed:
Remove unnecessary overrides

piotrbulinski committed:
Convert to async

piotrbulinski committed:
Fix and cleanup

piotrbulinski committed:
Remove duplicate device.update() from async_setup_entry

piotrbulinski committed:
Raise ConfigEntryNotReady on device.update timeout; update flexit_bacnet to 2.0.2

piotrbulinski committed:
Simplify config_flow.async_step_user logic

piotrbulinski committed:
Delete translations

piotrbulinski committed:
Remove redundant 'not device.is_valid()' check; remove unused _LOGGER from climate.py

piotrbulinski committed:
Cleanup async_setup_entry

piotrbulinski committed:
Sort _attr_ on FlexitClimateEntity

piotrbulinski committed:
Rename entity; cleanup unique ID; remove extra_state_attributes

piotrbulinski committed:
Move ventilation/preset mapping to constants file

piotrbulinski committed:
Remove control of aux_heat from HVAC mode control

piotrbulinski committed:
Simplify config flow

piotrbulinski committed:
Reuse HASS constants; improve error handling

piotrbulinski committed:
Improve error handling

piotrbulinski committed:
Re-use HASS constants for config

piotrbulinski committed:
Upgrade flexit_bacnet; remove unnecessary type casting

piotrbulinski committed:
Add device info, improve testing, cleanup

piotrbulinski committed:
Upgrade flexit_bacnet to 2.1.0; cleanup constants; fix DeviceInfo

lellky committed:
Added codeowner

frenck commented:
Small note, the upstream PR in the brands repo adds a brand for Flexit, this PR should add the brands definition as well to match that.

lellky committed:
Adds brand: Flexit

Kazhuu reviewed (approved):
LGTM!

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,174 @@
+            identifiers={
+                (DOMAIN, device.serial_number),
+            },
+            name="Ventilation",
```
Why don't we set the device name here? `self._device.device_name`

lellky replied:
Device name would then be "HvacFnct21y_A" (for my unit). Or should we try to get the model here instead. I.e. in my case that would be "Flexit Nordic S4".

joostlek replied:
As long as the user can identify it

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,174 @@
+        except (asyncio.exceptions.TimeoutError, ConnectionError, DecodingError) as exc:
+            raise HomeAssistantError from exc
+
+    @property
+    def is_aux_heat(self) -> bool:
+        """Return true if aux heater.
+
+        Requires ClimateEntityFeature.AUX_HEAT.
+        """
+        return self._device.electric_heater
+
+    async def async_turn_aux_heat_on(self) -> None:
+        """Turn auxiliary heater on."""
+        try:
+            await self._device.enable_electric_heater()
+        except (asyncio.exceptions.TimeoutError, ConnectionError, DecodingError) as exc:
+            raise HomeAssistantError from exc
+
+    async def async_turn_aux_heat_off(self) -> None:
+        """Turn auxiliary heater off."""
+        try:
+            await self._device.disable_electric_heater()
+        except (asyncio.exceptions.TimeoutError, ConnectionError, DecodingError) as exc:
+            raise HomeAssistantError from exc
```
I am no expert on climate entities, but have a read at https://github.com/home-assistant/architecture/discussions/932

lellky replied:
Removed

joostlek commented on homeassistant/components/flexit_bacnet/config_flow.py:
```diff
@@ -0,0 +1,78 @@
+)
+
+
+async def validate_input(data: dict[str, Any]) -> str:
```
This function is fairly small, I wouldn't mind if you inlined this

lellky replied:
Fixed

joostlek commented on tests/components/flexit_bacnet/test_config_flow.py:
```diff
@@ -0,0 +1,203 @@
+    return result["flow_id"]
+
+
+@pytest.fixture(autouse=True)
+def mock_serial_number():
+    """Mock serial number of the device."""
+    with patch(
+        "homeassistant.components.flexit_bacnet.config_flow.FlexitBACnet.serial_number",
+        property(lambda _: "0000-0001"),
+    ):
+        yield
+
+
+@pytest.fixture
+def mock_setup_entry():
+    """Mock setting up a config entry."""
+    with patch(
+        "homeassistant.components.flexit_bacnet.async_setup_entry", return_value=True
+    ) as setup_entry_mock:
+        yield setup_entry_mock
```
Maybe move these to conftest.py?

joostlek commented on tests/components/flexit_bacnet/test_config_flow.py:
```diff
@@ -0,0 +1,203 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_device(
```
Checkout `@pytest.mark.parametrize` you can easily throw different parameters at your test. checkout the Lastfm config flow for an example

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,174 @@
+            },
+            name="Ventilation",
+            manufacturer="Flexit",
+            model="Nordic",
+        )
```
```suggestion
            model="Nordic",
            serial_number=device.serial_number,
        )
```

joostlek commented on homeassistant/components/flexit_bacnet/config_flow.py:
```diff
@@ -0,0 +1,78 @@
+                await self.async_set_unique_id(info)
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(title=info, data=user_input)
```
I think the device name would be friendlier to use here

joostlek reviewed (changes requested):
This is a good first shot!

lellky committed:
Set the device name to the actual device name

Also use correct helper for DeviceInfo

lellky committed:
Adds serial number to device

lellky committed:
Moves mocks to conftest.py

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,175 @@
+        | ClimateEntityFeature.AUX_HEAT
+    )
+
+    _attr_target_temperature_step = 1.0
```
```suggestion
    _attr_target_temperature_step = PRECISION_WHOLE
```

lellky replied:
Fixed

lellky committed:
No name on entity

lellky committed:
Use const instead of number

lellky committed:
Refactors function to be inline code instead

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,172 @@
+            identifiers={
+                (DOMAIN, device.serial_number),
+            },
+            name=self._device.device_name,
```
```suggestion
            name=device.device_name,
```

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,172 @@
+        """Refresh unit state."""
+        await self._device.update()
+
+    _attr_name = None
```
Can be put on top

lellky replied:
Done

joostlek commented on homeassistant/components/flexit_bacnet/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
```
```suggestion
```

lellky replied:
Removed

lellky committed:
Moves attribute to top

lellky committed:
Use device, not self._device

lellky committed:
Removes unused class

lellky committed:
Removes possibility to turn on/off aux heat for now

https://github.com/home-assistant/architecture/discussions/932

Also removed is_aux_heat

joostlek commented on homeassistant/components/flexit_bacnet/climate.py:
```diff
@@ -0,0 +1,158 @@
+    _attr_supported_features = (
+        ClimateEntityFeature.PRESET_MODE
+        | ClimateEntityFeature.TARGET_TEMPERATURE
+        | ClimateEntityFeature.AUX_HEAT
```
```suggestion
```

lellky replied:
Removed

joostlek commented on tests/components/flexit_bacnet/conftest.py:
```diff
@@ -0,0 +1,49 @@
+    """Mock serial number of the device."""
+    with patch(
+        "homeassistant.components.flexit_bacnet.config_flow.FlexitBACnet.serial_number",
+        property(lambda _: "0000-0001"),
```
```suggestion
        return_value="0000-0001",
```

lellky replied:
Fixed

lellky committed:
Do not use lambda

lellky committed:
Fixes tests

lellky committed:
Adds test for already configured device

lellky committed:
Use parametrize for similar tests i.e. test_flow_fails

joostlek commented on tests/components/flexit_bacnet/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+
+from ...common import MockConfigEntry
```
```suggestion
from tests.common import MockConfigEntry
```

joostlek commented on tests/components/flexit_bacnet/test_config_flow.py:
```diff
@@ -0,0 +1,111 @@
+from homeassistant.data_entry_flow import FlowResultType
+
+from ...common import MockConfigEntry
+from .conftest import _patch_update
```
I don't think mypy allows you to import privates like this, maybe move the config flow fixtures back over (I know I told to move them there in the first place)

lellky replied:
Okay, will do!

lellky committed:
Fixes import

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

lellky committed:
Inline patch for update

lellky committed:
Combines patches
